opt/memo: improve zigzag join cost and selectivity estimation with multi-column stats

rytaft opened the pull request:
**opt: update seek and distribution cost of zigzag join to match scan**

Prior to this commit, the optimizer could prefer a zigzag join over a
scan even if they produced the same number of rows. This was because scans
always included the cost of at least one seek (involving random I/O) and
some distribution cost, while zigzag joins did not. This commit updates
the cost of zigzag joins to include seek and distribution costs so they
will never be chosen over scans unless they produce fewer rows.

This change is behind the setting `optimizer_use_improved_zigzag_join_costing`.

Release note (performance improvement): Added a new setting
`optimizer_use_improved_zigzag_join_costing`. When enabled, the cost of zigzag
joins is updated so they will be never be chosen over scans unless they
produce fewer rows. This change only matters if the setting `enable_zigzag_join`
is also true.

**opt/memo: improve selectivity estimation with multi-column stats**

This commit updates `correlationFromMultiColDistinctCounts` in `statisticsBuilder`
to use a tighter lower bound for the multi-column selectivity. This avoids
cases where we significantly over-estimate the selectivity of a multi-column
predicate.

Fixes https://github.com/cockroachdb/cockroach/issues/121397

Release note (performance improvement): Improved the selectivity estimation of
multi-column filters when the multi-column distinct count is high. This avoids
cases where we significantly over-estimate the selectivity of a multi-column
predicate and as a result can prevent the optimizer from choosing a bad query
plan.

**sql: add setting optimizer_use_improved_multi_column_selectivity_estimate**

Informs https://github.com/cockroachdb/cockroach/issues/121397

Release note (sql change): Added a setting
`optimizer_use_improved_multi_column_selectivity_estimate`, which if enabled,
causes the optimizer to use an improved selectivity estimate for multi-column
predicates. This setting will default to true on versions 24.2+, and false
on prior versions.

**opt: improve variable names in selectivityFromMultiColDistinctCounts**

This commit improves the variable names in
`selectivityFromMultiColDistinctCounts` in `statisticsBuilder` to be more
self-documenting.

Release note: None

rytaft committed:
opt/memo: improve selectivity estimation with multi-column stats

This commit updates correlationFromMultiColDistinctCounts in statisticsBuilder
to use a tighter lower bound for the multi-column selectivity. This avoids
cases where we significantly over-estimate the selectivity of a multi-column
predicate.

Fixes #121397

Release note (performance improvement): Improved the selectivity estimation of
multi-column filters when the multi-column distinct count is high. This avoids
cases where we significantly over-estimate the selectivity of a multi-column
predicate and as a result can prevent the optimizer from choosing a bad query
plan.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/120805)
<!-- Reviewable:end -->

rytaft committed:
sql: add setting optimizer_use_improved_multi_column_selectivity_estimate

Informs #121397

Release note (sql change): Added a setting
optimizer_use_improved_multi_column_selectivity_estimate, which if enabled,
causes the optimizer to use an improved selectivity estimate for multi-column
predicates. This setting will default to true on versions 24.2+, and false
on prior versions.

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @stevendanna)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuHalzPDcDVFXwpqNj4:-NuHalzPDcDVFXwpqNj5:b-d818i8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6961a00cc2d277c763926f10e136c264edfb6cff/pkg/bench/rttanalysis/testdata/benchmark_expectations#L106)):*
> ```
> 4,ORMQueries/pg_type
> 133,ORMQueries/prisma_column_descriptions
> 133,ORMQueries/prisma_column_descriptions_updated
> ```

@fqazi should I be concerned about this change?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
I guess the tradeoff here is that we'll consider multi-column correlations to be lower in general, so plans depending on high correlation might not be chosen? Does that sound right?

> Reviewed 7 of 18 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-NuRHZ2c4wldeNrHtorJ:bji8tz4)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4072 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuR6N-O-fzflFoYX7BG:-NuR6N-O-fzflFoYX7BH:b-ol9guw) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4072)):*
> ```Go
> 	// multiColSelectivity must be at least as large as singleColSelectivity,
> 	// since singleColSelectivity corresponds to equation (1).
> 	multiColSelectivity = props.MaxSelectivity(multiColSelectivity, singleColSelectivity)
> ```

Do I understand it correctly that the `lowerBound` in `correlationFromMultiColDistinctCounts` was basically `singleColSelectivity` before, and was the product of selectivities? And now the lower bound comes from the most selective single column?

Hmm, wait, I must be wrong because the most selective single column is `minLocalSel` which is the upper bound. :thinking: I'm a little confused here. How is `multiColSelectivityLowerBound` different from `minLocalSel`?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4072 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuR6N-O-fzflFoYX7BG:-NuVG6Gm3Xrche3qC2od:b-kpvixp) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4072)):*
Think of `multiColSelectivityLowerBound` as the minimum possible selectivity you could calculate for a distinct-count-based stat for a given predicate, without taking into account the actual multi-column stats provided. If the new distinct count is 1 (e.g., with a predicate like `x = 1 AND y = 1`) and there are 1000 rows in a table, the minimum selectivity would be 1/1000. 

> How is `multiColSelectivityLowerBound` different from `minLocalSel`?

The difference is in the denominator. The denominator for `minLocalSel` is the input distinct count of the most selective column, whereas the denominator for `multiColSelectivityLowerBound` is the input row count. The row count is the largest possible denominator for a distinct-count-based selectivity calculation, so therefore it represents a lower bound.  Technically we could get an even tighter lower bound by using `min(oldDistinctProduct, inputRowCount)`. (If this makes sense to you, maybe I'll go ahead and make that change...)


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Yes, this does help!

Also, not sure how relevant it is, but I noticed https://github.com/cockroachdb/cockroach/issues/121790 while playing around with this.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-NuaEJGgEtDczw0RFbFW:bux51tv)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 3999 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nuf6BpVDcT3KjkOSV7s:-NuRH0OrB_YlYSSmBdqG:b-9nakxv) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L3999)):*
> ```Go
> 	if !sb.evalCtx.SessionData().OptimizerUseMultiColStats {
> 		selectivity, selectivityUpperBound = sb.selectivityFromSingleColDistinctCounts(cols, e, s)
> 		return selectivity, selectivityUpperBound, selectivity
> ```

(Maybe this doesn't matter much, since we almost always have MultiColStats enabled, but random thought.)

Instead of considering the product of selectivities as the selectivity here, should we be calling that `selectivityLowerBound`, and be calculating selectivity as `selectivityLowerBound + (selectivityUpperBound - selectivityLowerBound) * (unknownCorrelation ^ (numCols - 1))` or something? In other words, does it always make sense to assume correlation = 0 when correlation is unknown? It seems like correlation = 0 is the most aggressive assumption we could make, because it will lead to the smallest row count estimates for multi-column conjunctions, and maybe we could be more conservative?
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4072 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuR6N-O-fzflFoYX7BG:-NuaFitk6DrKCXt505Lk:b-b68ib2) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4072)):*
I see, this helps (thank you)!

> The row count is the largest possible denominator for a distinct-count-based selectivity calculation, so therefore it represents a lower bound.

I see, this makes sense. So I guess another hand-wavy way of thinking of it is `singleColSelectivity` could be considered to be using `oldDistinctProduct` as denominator, but this could be larger than the actual input row count.

> Technically we could get an even tighter lower bound by using `min(oldDistinctProduct, inputRowCount)`. (If this makes sense to you, maybe I'll go ahead and make that change...)

Oh, right, you're saying the same thing! Yes, this makes sense now, go for it!
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4211 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuaItua9ryYWo94NFu_:-NuaItua9ryYWo94NFua:bndojwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4211)):*
> ```Go
> 	}
> 
> 	lowerBound, _ := sb.selectivityFromSingleColDistinctCounts(cols, e, s)
> ```

I might be saying this wrong, but is the reason you're keeping this lower bound because the input row count for selectivity calculation purposes will be the product of the left and right input row counts? And that will always be greater than the product of input distinct counts? :thinking:


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Great, thanks!

> Also, not sure how relevant it is, but I noticed [![](https://avatars.githubusercontent.com/u/32038551?s=60&v=4) opt: under-estimating join row count with multiple predicates leads to lookup join #121790](https://github.com/cockroachdb/cockroach/issues/121790) while playing around with this.

Might be, although the change I made below to the join stats estimation logic didn't seem to change much with our estimates for joins. :woman_shrugging: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nv2RTuA00v9hYMd-4EM:b4u0exr)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 3999 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nuf6BpVDcT3KjkOSV7s:-Nv2Rihj39ywKA9nugmv:b3kqo9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L3999)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

(Maybe this doesn't matter much, since we almost always have MultiColStats enabled, but random thought.)

Instead of considering the product of selectivities as the selectivity here, should we be calling that `selectivityLowerBound`, and be calculating selectivity as `selectivityLowerBound + (selectivityUpperBound - selectivityLowerBound) * (unknownCorrelation ^ (numCols - 1))` or something? In other words, does it always make sense to assume correlation = 0 when correlation is unknown? It seems like correlation = 0 is the most aggressive assumption we could make, because it will lead to the smallest row count estimates for multi-column conjunctions, and maybe we could be more conservative?
</blockquote></details>

I think this could work, but if it's ok with you, I'd like to keep this PR scoped to the issue I found with multi-column stats enabled. Also, I don't think we've seen a customer run into an issue here, so I'm not too motivated to change this (but please feel free to open a PR if you'd like, I'm not against this idea).
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4072 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuR6N-O-fzflFoYX7BG:-Nv2T2EfA9YhPlZpyQDP:boyexb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4072)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I see, this helps (thank you)!

> The row count is the largest possible denominator for a distinct-count-based selectivity calculation, so therefore it represents a lower bound.

I see, this makes sense. So I guess another hand-wavy way of thinking of it is `singleColSelectivity` could be considered to be using `oldDistinctProduct` as denominator, but this could be larger than the actual input row count.

> Technically we could get an even tighter lower bound by using `min(oldDistinctProduct, inputRowCount)`. (If this makes sense to you, maybe I'll go ahead and make that change...)

Oh, right, you're saying the same thing! Yes, this makes sense now, go for it!
</blockquote></details>

Done. Thanks!
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4211 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuaItua9ryYWo94NFu_:-Nv2TSUJ9zCJtZzU5iZE:bxvghqy) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4211)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I might be saying this wrong, but is the reason you're keeping this lower bound because the input row count for selectivity calculation purposes will be the product of the left and right input row counts? And that will always be greater than the product of input distinct counts? :thinking:
</blockquote></details>

Good catch! I just didn't actually think through this case. I think we can use the new lower bound here. Done. (although it doesn't seem to change the estimates too much for our existing test cases)


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
I have some questions about a couple of the plan changes.

Purely cosmetic, but I wonder if we're going to get complaints about estimates of 0.333 being rounded down to 0 in EXPLAIN output? Would it be worth changing this line to use `math.Ceil` instead of `math.Round`?

https://github.com/cockroachdb/cockroach/blob/f9ab8e3c2ad0d38f0fbcea66e483344df5fcc6db/pkg/sql/opt/exec/explain/emit.go#L526

(If you don't want to, I'm willing to do this in another PR.)

> Reviewed 28 of 44 files at r1, 5 of 18 files at r3, 13 of 14 files at r4, 2 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nv46yDS2CmpSKqzIcm2:b96a29t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @mgartner, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 3999 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nuf6BpVDcT3KjkOSV7s:-Nv3nLuBEUcctcvnnaJh:b-81nlyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L3999)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think this could work, but if it's ok with you, I'd like to keep this PR scoped to the issue I found with multi-column stats enabled. Also, I don't think we've seen a customer run into an issue here, so I'm not too motivated to change this (but please feel free to open a PR if you'd like, I'm not against this idea).
</blockquote></details>

SGTM.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4233 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4-Vi41_Zjs-a0aTKo:-Nv4-Vi41_Zjs-a0aTKp:b-nw6vzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/memo/statistics_builder.go#L4233)):*
> ```Go
> 	upperBound.Multiply(upperBoundRight)
> 	lowerBound := lowerBoundLeft
> 	lowerBound.Multiply(lowerBoundRight)
> ```

I think this needs the `sb.evalCtx.SessionData().OptimizerUseImprovedMultiColumnSelectivityEstimate` check, right?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/forecast1401` line 18488 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv45jL_1vt1sfadDQyC:-Nv45jLaATUu1Es6gFyB:b-swfton) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/exec/execbuilder/testdata/forecast1401#L18488)):*
> ```
>         │
>         └── • scan
>               estimated row count: 20 (<0.01% of the table; stats collected <hidden> ago; using stats forecast)
> ```

Is the limit hint not pushed down anymore? :thinking: Oh, we don't print it if equal to the row count. Interesting.
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-Nv4BcSg7zvZ7FFSimz-:b7qakwh) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
> ```
> --
> optimal plan:
> inner-join (zigzag t1@_hyp_1 t1@_hyp_2)
> ```

Do we have zz join turned on in the indexrec tests? :thinking:
___
*[`pkg/sql/opt/memo/testdata/stats/select` line 3461 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4EANAEK5uUXkeDmK9:-Nv4EANAEK5uUXkeDmKA:b-za6rua) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbae23207afa032ffe4af1b6158c3d56661c3189/pkg/sql/opt/memo/testdata/stats/select#L3461)):*
> ```
> ----
> 
> # Zig-zag join should not be picked due to high selectivity.
> ```

Is this a regression?
___
*[`pkg/sql/opt/xform/testdata/coster/scan` line 154 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4HDddAEX90-qFALuA:-Nv4HDdeDUzk2bxUv5SK:b-6r3sb7) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/xform/testdata/coster/scan#L154)):*
> ```
>  │    │    ├── key: (1)
>  │    │    ├── fd: (1)-->(2-4)
>  │    │    └── limit hint: 400.00
> ```

Why does the limit hint change like this?
___
*[`pkg/sql/opt/xform/testdata/external/tpcc-no-stats` line 958 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4Iq4ICwfLwuGRgzuy:-Nv4Iq4ICwfLwuGRgzuz:b6cct95) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/xform/testdata/external/tpcc-no-stats#L958)):*
> ```
>       │         ├── scan customer@customer_idx
>       │         │    ├── columns: c_id:24!null c_d_id:25!null c_w_id:26!null c_first:27 c_last:29
>       │         │    ├── constraint: /26/25/29/27/24: [/10/1 - /10/10]
> ```

This seems like a regression to me, since before we were constraining on all three columns, and now we're constraining on two and using a index join.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 37 of 44 files at r1, 18 of 18 files at r3, 13 of 14 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @michae2, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCReiXDLJO2tQ9lI8R:-NvCReiXDLJO2tQ9lI8S:b2j1lub) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4122)):*
> ```Go
> 
> 	// multiColSelectivityLowerBound is the minimum multi-column selectivity, which
> 	// is the minimum possible new multi-col distinct count divided by the old row
> ```

Why does this say "minimum possible new multi-col distinct count" if `maxNewDistinct` is used? Is it because `maxNewDistinct` represents `min_value = max(new_distinct(x), new_distinct(y))`? It looks like "maximum" describes `min(inputStats.RowCount, oldDistinctProduct)` which is also confusing.
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-NvCT7eBBfszCSMchz3h:balstk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Do we have zz join turned on in the indexrec tests? :thinking:
</blockquote></details>

Even with it turned on, it's odd that a zigzag join beats a single value scan of an index. Even with zigzags disabled by default, I'm a little worried there could be a deeper issue causing this.
___
*[`pkg/sql/opt/testutils/opttester/testdata/use-multi-col-stats` line 38 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCel5U6O8f0v2xfiCE:-NvCel5U6O8f0v2xfiCF:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/testutils/opttester/testdata/use-multi-col-stats#L38)):*
> ```
>  ├── scan rides
>  │    ├── columns: id:1(uuid!null) city:2(varchar!null) vehicle_city:3(varchar) rider_id:4(uuid) vehicle_id:5(uuid) start_address:6(varchar) end_address:7(varchar) start_time:8(timestamp) end_time:9(timestamp) revenue:10(decimal)
>  │    ├── constraint: /2/1: [/'rome' - /'rome']
> ```

Nice!
___
*[`pkg/sql/opt/xform/testdata/coster/scan` line 154 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4HDddAEX90-qFALuA:-NvCeqWT-t-lCRDiUuO3:bnkvnnu) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/xform/testdata/coster/scan#L154)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Why does the limit hint change like this?
</blockquote></details>

The limit hint is determined from the ratio of the select's input row count to output row count. So this limit hint went up because the select's row count went down. Before the change 407/450 input rows were estimated to pass the filter, so we only needed to scan a few more than 20 rows to find 20 matches. Now that 22/450 input rows pass the filter, we have to scan more rows to find 20.
___
*[`pkg/sql/opt/xform/testdata/rules/insert` line 548 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCfxnSEzEGXc4dF04i:-NvCfxnSEzEGXc4dF04j:b-6gei2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/insert#L548)):*
> ```
>                 └── scan t@t_b_a_idx
>                      ├── columns: t.k:70!null t.a:72!null t.b:73!null
>                      ├── constraint: /73/72/70: [/2/10 - /2/10]
> ```

This seems like a regression, but I'm not sure. I don't understand why the `INDEX (b, a)` even exists in this table.
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 2666 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCgWnN3DH1J5QOMKye:-NvCgWnN3DH1J5QOMKyf:bwllldg) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/limit#L2666)):*
> ```
>  │    │    │    │         └── (((b:9 <= 11) AND (c:10 < 50)) OR ((b:9 = 11) AND (c:10 = 50))) OR ((b:9 >= 11) AND (c:10 > 50)) [outer=(9,10), constraints=(/9: (/NULL - ]; /10: (/NULL - /49] [/50 - /50] [/51 - ])]
>  │    │    │    └── 1
>  │    │    └── sort
> ```

This change is certainly surprising...


<!-- Sent from Reviewable.io -->

rytaft committed:
opt: update seek and distribution cost of zigzag join to match scan

Prior to this commit, the optimizer could prefer a zigzag join over a
scan even if they produced the same number of rows. This was because scans
always included the cost of at least one seek (involving random I/O) and
some distribution cost, while zigzag joins did not. This commit updates
the cost of zigzag joins to include seek and distribution costs so they
will never be chosen over scans unless they produce fewer rows.

This change is behind the setting optimizer_use_improved_zigzag_join_costing.

Release note (performance improvement): Added a new setting
optimizer_use_improved_zigzag_join_costing. When enabled, the cost of zigzag
joins is updated so they will be never be chosen over scans unless they
produce fewer rows. This change only matters if the setting enable_zigzag_join
is also true.

rytaft committed:
opt: improve variable names in selectivityFromMultiColDistinctCounts

This commit improves the variable names in
selectivityFromMultiColDistinctCounts in statisticsBuilder to be more
self-documenting.

Release note: None

rytaft reviewed (commented):
Thanks for the thorough reviews!

> Purely cosmetic, but I wonder if we're going to get complaints about estimates of 0.333 being rounded down to 0 in EXPLAIN output? Would it be worth changing this line to use `math.Ceil` instead of `math.Round`?

I think we have a lot of zeros already in EXPLAIN output, there were just a few more added in this PR. I'd be in favor of keeping the logic to use `math.Round` as-is. If someone does complain in the future, maybe we could always show the first significant digit?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-NvYBLiBBPPPzkA_QU8h:b-osbwys)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @fqazi, @mgartner, @michae2, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCReiXDLJO2tQ9lI8R:-NvY4zCP9S6C4oNeyhOf:bvngepc) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4122)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Why does this say "minimum possible new multi-col distinct count" if `maxNewDistinct` is used? Is it because `maxNewDistinct` represents `min_value = max(new_distinct(x), new_distinct(y))`? It looks like "maximum" describes `min(inputStats.RowCount, oldDistinctProduct)` which is also confusing.
</blockquote></details>

Yea, I agree it's confusing. I added a new commit that updates all the variable names -- let me know if you think it's less confusing now.
___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4233 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4-Vi41_Zjs-a0aTKo:-NvYAB4t9M8BInZBHIJ1:b-o7vuhc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/memo/statistics_builder.go#L4233)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I think this needs the `sb.evalCtx.SessionData().OptimizerUseImprovedMultiColumnSelectivityEstimate` check, right?
</blockquote></details>

Good catch! Done.
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-Nvd-LxaDGWc5xW91czC:b-52uiig) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Even with it turned on, it's odd that a zigzag join beats a single value scan of an index. Even with zigzags disabled by default, I'm a little worried there could be a deeper issue causing this.
</blockquote></details>

I looked into this and found that this was an issue with zigzag join costing. I've added another commit to this PR that fixes the costing issue. One question I have is whether we should backport that commit and if so, whether I should add another session setting for it (or if you'd rather me open a separate PR).
___
*[`pkg/sql/opt/memo/testdata/stats/select` line 3461 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4EANAEK5uUXkeDmK9:-Nvd-gMN6yPojHOLitDt:bb7228t) ([raw file](https://github.com/cockroachdb/cockroach/blob/fbae23207afa032ffe4af1b6158c3d56661c3189/pkg/sql/opt/memo/testdata/stats/select#L3461)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is this a regression?
</blockquote></details>

No-- with the new stats logic, the row count estimate reduced from 29371 to 12498, which caused zigzag join to be selected. I checked the support issue, and the relevant query is the next one using `count(c)`, which still avoids the zigzag join. I think this is an acceptable change.
___
*[`pkg/sql/opt/xform/testdata/external/tpcc-no-stats` line 958 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4Iq4ICwfLwuGRgzuy:-NvZufwU6bqRd77o46n3:b1duivw) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/xform/testdata/external/tpcc-no-stats#L958)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

This seems like a regression to me, since before we were constraining on all three columns, and now we're constraining on two and using a index join.
</blockquote></details>

The reason is that we have reduced the row count estimate of the three predicates from 9 rows down to 0.1 rows. The cost of seeking to each span in the 3-column constraint outweighs the cost of scanning one span in the 2-column constraint. I'm not too concerned about this one, since it's TPC-C with no stats. The plans for TPC-C with stats have not changed.
___
*[`pkg/sql/opt/xform/testdata/rules/insert` line 548 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCfxnSEzEGXc4dF04i:-Nvd0JSK1Q3WwTJDtZfE:b9edfzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/insert#L548)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This seems like a regression, but I'm not sure. I don't understand why the `INDEX (b, a)` even exists in this table.
</blockquote></details>

 I looked into this and I think this is fine since it's scanning the same number of rows (there are only two regions -- east and west). The difference is that with the lower row count estimate, it was determined to be cheaper to perform an index join using `t_b_a_idx` rather than seek to two spans in `t_r_a_b_idx`.
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 2666 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCgWnN3DH1J5QOMKye:-NvdVWTB-Y8SwcrfFYzM:b4hnod0) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/limit#L2666)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This change is certainly surprising...
</blockquote></details>

Yea, doesn't seem worse, though, just different. But let me know if you think I should dig deeper and I can try to figure out why this changed.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo @michae2's comments

> Reviewed 61 of 61 files at r6, 47 of 47 files at r7, 17 of 17 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nvlqfv45GhRV9nSR7wL:boy1bxc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @fqazi, @michae2, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4122 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCReiXDLJO2tQ9lI8R:-NvllIl3AADhdZyCNBhq:b-qbxl94) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/memo/statistics_builder.go#L4122)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Yea, I agree it's confusing. I added a new commit that updates all the variable names -- let me know if you think it's less confusing now.
</blockquote></details>

Looks great, thanks!
___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuHalzPDcDVFXwpqNj4:-NvlkmheD9G0OG3VypUl:bmyqsy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6961a00cc2d277c763926f10e136c264edfb6cff/pkg/bench/rttanalysis/testdata/benchmark_expectations#L106)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

@fqazi should I be concerned about this change?
</blockquote></details>

Friendly ping @fqazi.
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-NvlnjkMAJ-3tGOefdm-:bhpj6e1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I looked into this and found that this was an issue with zigzag join costing. I've added another commit to this PR that fixes the costing issue. One question I have is whether we should backport that commit and if so, whether I should add another session setting for it (or if you'd rather me open a separate PR).
</blockquote></details>

My main concern is backporting to 23.1 where zig-zag joins are enabled by default, and turning on multi-column stats improvements makes zigzag joins more likely to be chosen. I'd be fine backporting the zigzag join changes behind the same flag in 23.1 to prevent that. But I don't necessarily think we should keep the zigzag costing fix behind the same flag in 24.1+. Would that be weird to have the flag do slightly different things across versions?
___
*[`pkg/sql/opt/xform/testdata/rules/insert` line 548 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCfxnSEzEGXc4dF04i:-Nvlpmbf33GrkYOJOQHi:bzdbkw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/insert#L548)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I looked into this and I think this is fine since it's scanning the same number of rows (there are only two regions -- east and west). The difference is that with the lower row count estimate, it was determined to be cheaper to perform an index join using `t_b_a_idx` rather than seek to two spans in `t_r_a_b_idx`.
</blockquote></details>

Hmm interesting. The index-join seems completely unnecessary to me—the fetch column `r:71` isn't needed. Fixing that would probably require something like #67731.

The test case does seem contrived though—if we have an `UNIQUE WITHOUT INDEX (a, b)` and an `INDEX(b, a)`, it'd be better to have a just a `UNIQUE INDEX (b, a)`. So I don't think this is worth worrying about, but maybe something to clean up in the future to avoid confusion.
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 2666 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCgWnN3DH1J5QOMKye:-Nvlm68TCqbzhPyGNior:b-d9tcsf) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/limit#L2666)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Yea, doesn't seem worse, though, just different. But let me know if you think I should dig deeper and I can try to figure out why this changed.
</blockquote></details>

Looking at it again, I actually think it's an improvement because the new plan sorts fewer rows (only those on the RHS of the union-all) and the scan on LHS of the union-all should only scan a single batch before satisfying the limit=1, because the select filters should always be true for the scanned rows. (If the optimizer detected that the select filters are true for all input rows then the limit could be pushed into the scan as a hard limit, but I believe the problem is due to limitations of the constraint library in representing disjunctions and not easy to solve). 


<!-- Sent from Reviewable.io -->

fqazi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, @rytaft, and @stevendanna)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuHalzPDcDVFXwpqNj4:-NvqnZEeAYif-aUOrWGm:b-p30vjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6961a00cc2d277c763926f10e136c264edfb6cff/pkg/bench/rttanalysis/testdata/benchmark_expectations#L106)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Friendly ping @fqazi.
</blockquote></details>

Yeah thats gonna cause a pretty bad regression for Prisma. We were trying to force a plan that avoids scanning the partial secondary index for them in pg_class. See: https://github.com/cockroachdb/cockroach/issues/111047.

We need to either avoid regressing that query or have to modify the upstream again to get a better plan for this query, since its going to have a severe impact on multiregion with the round trips.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi, @mgartner, @michae2, and @stevendanna)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuHalzPDcDVFXwpqNj4:-Nvr0sg02Y_7si2_3v41:bjyjlh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6961a00cc2d277c763926f10e136c264edfb6cff/pkg/bench/rttanalysis/testdata/benchmark_expectations#L106)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Yeah thats gonna cause a pretty bad regression for Prisma. We were trying to force a plan that avoids scanning the partial secondary index for them in pg_class. See: https://github.com/cockroachdb/cockroach/issues/111047.

We need to either avoid regressing that query or have to modify the upstream again to get a better plan for this query, since its going to have a severe impact on multiregion with the round trips.
</blockquote></details>

Thanks for the context! Let me see what I can do...
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-Nvr-iW11Ve_MolDgF5M:b5kvayb) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
> My main concern is backporting to 23.1 where zig-zag joins are enabled by default, and turning on multi-column stats improvements makes zigzag joins more likely to be chosen.

Good call.

> ... Would that be weird to have the flag do slightly different things across versions?

Yea I think so.

I added a new flag `optimizer_use_improved_zigzag_join_costing`. I think we should just backport both settings and tell customers who want to use `optimizer_use_improved_multi_column_selectivity_estimate` that they should also use `optimizer_use_improved_zigzag_join_costing`.
___
*[`pkg/sql/opt/xform/testdata/rules/insert` line 548 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCfxnSEzEGXc4dF04i:-Nvr0KMO0WgI1lRBiib1:b-8fx2u7) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/insert#L548)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Hmm interesting. The index-join seems completely unnecessary to me—the fetch column `r:71` isn't needed. Fixing that would probably require something like #67731.

The test case does seem contrived though—if we have an `UNIQUE WITHOUT INDEX (a, b)` and an `INDEX(b, a)`, it'd be better to have a just a `UNIQUE INDEX (b, a)`. So I don't think this is worth worrying about, but maybe something to clean up in the future to avoid confusion.
</blockquote></details>

:+1: 
___
*[`pkg/sql/opt/xform/testdata/rules/limit` line 2666 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvCgWnN3DH1J5QOMKye:-Nvr0QuuEqMp4_j6J11E:bg7yuj) ([raw file](https://github.com/cockroachdb/cockroach/blob/326083bf8792a2a3818cda35d465778e614810e0/pkg/sql/opt/xform/testdata/rules/limit#L2666)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looking at it again, I actually think it's an improvement because the new plan sorts fewer rows (only those on the RHS of the union-all) and the scan on LHS of the union-all should only scan a single batch before satisfying the limit=1, because the select filters should always be true for the scanned rows. (If the optimizer detected that the select filters are true for all input rows then the limit could be pushed into the scan as a hard limit, but I believe the problem is due to limitations of the constraint library in representing disjunctions and not easy to solve).
</blockquote></details>

Nice, thanks for the analysis.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> except for the `prisma_column_descriptions_updated` regression. Nice work!

> Reviewed 1 of 62 files at r10, 61 of 61 files at r14, 47 of 47 files at r15, 16 of 16 files at r16, 1 of 1 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-NvsCQ92BcFK5GoeKimR:b-xzg0n5)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi, @mgartner, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-NvrpLY00GKJA81z1tKp:b-393lwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

> My main concern is backporting to 23.1 where zig-zag joins are enabled by default, and turning on multi-column stats improvements makes zigzag joins more likely to be chosen.

Good call.

> ... Would that be weird to have the flag do slightly different things across versions?

Yea I think so.

I added a new flag `optimizer_use_improved_zigzag_join_costing`. I think we should just backport both settings and tell customers who want to use `optimizer_use_improved_multi_column_selectivity_estimate` that they should also use `optimizer_use_improved_zigzag_join_costing`.
</blockquote></details>

Nice fix! Would it make sense to backport with `optimizer_use_improved_zigzag_join_costing` enabled by default? Then we wouldn't need special instructions for `optimizer_use_improved_multi_column_selectivity_estimate`, and we could always disable `optimizer_use_improved_zigzag_join_costing` if someone wasn't getting the zigzag join they needed (which seems unlikely).
___
*[`pkg/sql/opt/xform/testdata/external/trading` line 927 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NvsAHDv4n4Vv-pCxLDg:-NvsAHDv4n4Vv-pCxLDh:b-83sjoh) ([raw file](https://github.com/cockroachdb/cockroach/blob/0a06b1291510538aaee86f40679a4d79344e4dbb/pkg/sql/opt/xform/testdata/external/trading#L927)):*
> ```
> #   4. The row count estimate for the constrained scan of transactions is too
> #      large, causing us to choose the incorrect join algorithm (we should
> #      choose a lookup join). Collecting small histograms on all columns would
> ```

Nice!


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
TFTRs!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nvt8wJn8pVFaDMZxk4w:bz5mair)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi, @mgartner, @michae2, and @stevendanna)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuHalzPDcDVFXwpqNj4:-Nvt6Xsb-PJCyIA3ndQk:boqqpya) ([raw file](https://github.com/cockroachdb/cockroach/blob/6961a00cc2d277c763926f10e136c264edfb6cff/pkg/bench/rttanalysis/testdata/benchmark_expectations#L106)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Thanks for the context! Let me see what I can do...
</blockquote></details>

I've investigated this and I think it's going to be hard to fix in this PR, since we currently don't have stats on virtual tables. Even if we did, the optimizer doesn't have any insight into how many real rows must be read to produce a certain number of rows from the virtual table. I think the right way to handle this is to add a hint on this query to force the optimizer to use the "primary" index of the `pg_class` virtual table. We don't currently allow index hints on virtual tables, but I don't think it would be too hard to add. We would need to modify the Prisma upstream again, but the advantage is that the index hint would force the correct plan even if the stats code changes in the future.

Does this seem like a reasonable plan, @fqazi? If so, I'll merge this PR as-is since we have a customer waiting on the fix. I'll backport the PR with the settings disabled, so existing Prisma users won't run into this. I'll be sure to get a fix to the Prisma query merged before 24.2 (I can open an issue to track it if you're ok with the plan).
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-Nvt0bNeBy-bRHK09TCB:bmnx9bg) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Nice fix! Would it make sense to backport with `optimizer_use_improved_zigzag_join_costing` enabled by default? Then we wouldn't need special instructions for `optimizer_use_improved_multi_column_selectivity_estimate`, and we could always disable `optimizer_use_improved_zigzag_join_costing` if someone wasn't getting the zigzag join they needed (which seems unlikely).
</blockquote></details>

I think we should avoid backporting any changes that could cause unexpected plan changes in a patch release (even if it seems unlikely to be bad, there's always a chance it could be). Customers are not likely to use `optimizer_use_improved_multi_column_selectivity_estimate` without our guidance anyway, so I don't think it's too bad to tell them to also use `optimizer_use_improved_zigzag_join_costing`.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
I was able to fix the `prisma_column_descriptions_updated` regression with a hack that takes advantage of the fact that stats are not available for virtual tables. PTAL @michae2 @mgartner 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nvtc8G05cZ9nuoghDSI:bji8wk3)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @fqazi, @mgartner, @michae2, and @stevendanna)

___
*[`pkg/bench/rttanalysis/testdata/benchmark_expectations` line 106 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-NuHalzPDcDVFXwpqNj4:-Nvt_wwJEAz6Q16dMZtO:b-78oced) ([raw file](https://github.com/cockroachdb/cockroach/blob/6961a00cc2d277c763926f10e136c264edfb6cff/pkg/bench/rttanalysis/testdata/benchmark_expectations#L106)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I've investigated this and I think it's going to be hard to fix in this PR, since we currently don't have stats on virtual tables. Even if we did, the optimizer doesn't have any insight into how many real rows must be read to produce a certain number of rows from the virtual table. I think the right way to handle this is to add a hint on this query to force the optimizer to use the "primary" index of the `pg_class` virtual table. We don't currently allow index hints on virtual tables, but I don't think it would be too hard to add. We would need to modify the Prisma upstream again, but the advantage is that the index hint would force the correct plan even if the stats code changes in the future.

Does this seem like a reasonable plan, @fqazi? If so, I'll merge this PR as-is since we have a customer waiting on the fix. I'll backport the PR with the settings disabled, so existing Prisma users won't run into this. I'll be sure to get a fix to the Prisma query merged before 24.2 (I can open an issue to track it if you're ok with the plan).
</blockquote></details>

Actually, I have a hack to fix this. Assuming the reviewers approve, I think we can prevent this regression for now.


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 1 of 62 files at r10, 1 of 61 files at r14, 60 of 60 files at r18, 51 of 51 files at r19, 11 of 47 files at r20.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nw5bGtM5CQCWPCQXNmp:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @fqazi, @michae2, and @stevendanna)

___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-Nw5_WpK4yGwpPOBoDCy:b-81nlyp) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I think we should avoid backporting any changes that could cause unexpected plan changes in a patch release (even if it seems unlikely to be bad, there's always a chance it could be). Customers are not likely to use `optimizer_use_improved_multi_column_selectivity_estimate` without our guidance anyway, so I don't think it's too bad to tell them to also use `optimizer_use_improved_zigzag_join_costing`.
</blockquote></details>

SGTM!


<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks! Friendly ping, @michae2 





<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Seems fine :slightly_smiling_face:

> Reviewed 60 of 60 files at r18, 51 of 51 files at r19, 47 of 47 files at r20, 1 of 26 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nw6pb_b7FbWBbda6YwA:bc1u0la)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @fqazi, @mgartner, @rytaft, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4280 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nw6osbBBYa6_c7FjhL4:-Nw6osbBBYa6_c7FjhL5:bb0ijjz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d033fb10382578ffb01cad1dac913118bea55122/pkg/sql/opt/memo/statistics_builder.go#L4280)):*
> ```Go
> 	// Check s.Available since we don't want to change plans where the stats were
> 	// not available anyway.
> 	if !sb.evalCtx.SessionData().OptimizerUseImprovedMultiColumnSelectivityEstimate || !s.Available {
> ```

Maybe another way to frame this is: `OptimizerUseImprovedMultiColumnSelectivityEstimate` tends to cause some hash join plans to turn into lookup join plans, and when we're operating without stats, that bias is more risky. That doesn't seem like too much of a hack.
___
*[`pkg/sql/opt/indexrec/testdata/index` line 637 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nv4BcSg7zvZ7FFSimyz:-Nw6hBeH-56oM3aRbtEA:bn1wz0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/b6800175f020e85f15cd6e17f4a87b74c4596a6a/pkg/sql/opt/indexrec/testdata/index#L637)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

SGTM!
</blockquote></details>

:thumbsup:


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Thanks all!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-:-Nw7yoikAyJZtV62fGBi:bbdmny1)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @fqazi, @michae2, and @stevendanna)

___
*[`pkg/sql/opt/memo/statistics_builder.go` line 4280 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/120805#-Nw6osbBBYa6_c7FjhL4:-Nw7yZEs3YjMzMfIl0dZ:b-igfwqu) ([raw file](https://github.com/cockroachdb/cockroach/blob/d033fb10382578ffb01cad1dac913118bea55122/pkg/sql/opt/memo/statistics_builder.go#L4280)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Maybe another way to frame this is: `OptimizerUseImprovedMultiColumnSelectivityEstimate` tends to cause some hash join plans to turn into lookup join plans, and when we're operating without stats, that bias is more risky. That doesn't seem like too much of a hack.
</blockquote></details>

Nice, thanks. I updated the comment with that language.


<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+





<!-- Sent from Reviewable.io -->

rytaft commented:
blathers backport staging-v23.1.20
